examples/forward_backward.py

```diff
@@ -0,0 +1,194 @@
+    for trans in factors[:0:-1]:
+        beta = prod_op(trans(**curr_to_drop), beta(**prev_to_drop)).reduce(
+            sum_op, reduce_vars
+        )
+        betas.append(beta)
+    else:
+        init = factors[0]
+        Z = prod_op(init(**curr_to_drop), beta(**prev_to_drop)).reduce(
+            sum_op, reduce_vars
+        )
+    betas.reverse()
+
+    # Forward algorithm
+    # p(y[0], x[0])
+    alpha = factors[0]
+    alphas = [alpha]
+    # p(x[0] | Y)
+    marginal = prod_op(factors[0], betas[0](**{"x_prev": "x_curr"})) / Z
+    marginals = [marginal]
+    # inductive steps
+    for trans, beta in zip(factors[1:], betas[1:]):
+        # p(y[0:t], x[t-1], x[t])
+        new_term = prod_op(alpha(**{"x_curr": "x_prev"}), trans)
+        # p(y[0:t], x[t])
+        alpha = new_term.reduce(sum_op, frozenset({"x_prev"}))
+        alphas.append(alpha)
+        # p(x[t-1], x[t] | Y)
+        marginal = prod_op(new_term, beta(**{"x_prev": "x_curr"})) / Z
+        marginals.append(marginal)
+
+    return marginals
+
+
+def main(args):
+    """
+    Compute marginal probabilities p(x[t], x[t-1] | Y) for an HMM:
+
+    x[0] -> ... -> x[t-1] -> x[t] -> ... -> x[T]
+     |              |         |             |
+     v              v         v             v
+    y[0]           y[t-1]    y[t]           y[T]
+
+    alpha[t] = alpha[t-1] @ p(y[t], x[t] | x[t-1])
+    beta[t-1] = p(y[t], x[t] | x[t-1]) @ beta[t]
+
+    dlog_Z / dlog_p(y[t], x[t] | x[t-1]) =
+    alpha[t-1] * beta[t] * p(y[t], x[t] | x[t-1]) / Z =
+    p(x[t], x[t-1] | Y)
+
+    **References:**
+
+    [1] Jason Eisner (2016)
+        "Inside-Outside and Forward-Backward Algorithms Are Just Backprop
+        (Tutorial Paper)"
+        https://www.cs.jhu.edu/~jason/papers/eisner.spnlp16.pdf
+    [2] Zhifei Li and Jason Eisner (2009)
+        "First- and Second-Order Expectation Semirings
+        with Applications to Minimum-Risk Training on Translation Forests"
+        http://www.cs.jhu.edu/~zfli/pubs/semiring_translation_zhifei_emnlp09.pdf
+    """
+    funsor.set_backend("torch")
+
+    # transition and emission probabilities
+    init = random_tensor(OrderedDict([("x_curr", Bint[args.hidden_dim])]))
+    log_factors = [init]
+    for t in range(args.time_steps - 1):
+        log_factors.append(
+            random_tensor(
+                OrderedDict(
+                    [
+                        ("x_prev", Bint[args.hidden_dim]),
+                        ("x_curr", Bint[args.hidden_dim]),
+                    ]
+                )
+            )
+        )
+    factors = [f.exp() for f in log_factors]
+
+    # Compute marginal probabilities using the forward-backward algorithm
+    marginals = forward_backward_algorithm(
+        ops.add, ops.mul, factors, {"x_prev": "x_curr"}
+    )
+    # Compute marginal probabilities using backpropagation
+    with AdjointTape() as tape:
+        log_Z = forward_algorithm(ops.add, ops.mul, factors, {"x_prev": "x_curr"})
```

Comment:
This works but I had to use `.log()` in `forward_algorithm` and register the adjoint of `LogOp`. Ideally I would like to run the forward algorithm with `log_factors` and use `ops.logaddexp` and `ops.add`. Then I think the derivative should work out as shown on line 130. However, at this moment it is not clear to me how `adjoint` module works with `ops.logaddexp` and `ops.add` and what are `adj_redop` and `adj_binop` are.